go/types/set_iterator.go

```diff
@@ -0,0 +1,208 @@
+		return si.cursor.current().(Value)
+	}
+	return nil
+}
+
+func (si *setIterator) AdvanceTo(v Value) Value {
+	d.Chk.NotNil(v, "setIterator.AdvanceTo() called with nil value")
+	first := false
+	if si.cursor == nil {
+		first = true
+		si.cursor, _ = si.s.getCursorAtValue(v)
+	}
+
+	if !si.cursor.valid() {
+		return nil
+	}
+
+	curValue := si.cursor.current().(Value)
+	if first {
+		return curValue
+	}
+
+	if v.Less(curValue) {
+		return si.Next()
+	}
+
+	si.cursor, _ = si.s.getCursorAtValue(v)
+	if si.cursor.valid() {
+		return si.cursor.current().(Value)
+	}
+	return nil
+}
+
+// iterState contains iterator and it's current value
+type iterState struct {
+	i SetIterator
+	v Value
+}
+
+func (st *iterState) Next() Value {
+	if st.v == nil {
+		return nil
+	}
+	v := st.v
+	st.v = st.i.Next()
+	return v
+}
+
+func (st *iterState) AdvanceTo(v Value) Value {
+	if st.v == nil || v == nil {
+		st.v = nil
+		return nil
+	}
+	st.v = st.i.AdvanceTo(v)
+	return st.v
+}
+
+type UnionIterator struct {
+	aState iterState
+	bState iterState
+}
+
+func NewUnionIterator(iterA, iterB SetIterator) SetIterator {
+	d.Chk.NotNil(iterA)
+	d.Chk.NotNil(iterB)
+	a := iterState{i: iterA, v: iterA.Next()}
+	b := iterState{i: iterB, v: iterB.Next()}
+	return &UnionIterator{aState: a, bState: b}
+}
+
+func (u *UnionIterator) Next() Value {
+	switch compareValue(u.aState.v, u.bState.v) {
+	case -1:
+		return u.aState.Next()
+	case 0:
+		u.aState.Next()
+		return u.bState.Next()
+	case 1:
+		return u.bState.Next()
+	}
+
+	return nil
+}
+
+func (u *UnionIterator) AdvanceTo(v Value) Value {
+	d.Chk.NotNil(v)
+	didAdvance := false
+	if u.aState.v != nil && u.aState.v.Less(v) {
+		didAdvance = true
+		u.aState.AdvanceTo(v)
+	}
+	if u.bState.v != nil && u.bState.v.Less(v) {
+		didAdvance = true
+		u.bState.AdvanceTo(v)
+	}
+	if !didAdvance {
+		return u.Next()
+	}
+	if u.aState.v == nil {
+		return u.bState.Next()
+	}
+	if u.bState.v == nil {
+		return u.aState.Next()
+	}
+	if u.aState.v.Less(u.bState.v) {
+		return u.aState.Next()
+	}
+	if u.aState.v.Equals(u.bState.v) {
+		u.aState.Next()
+		return u.bState.Next()
+	}
+	return u.bState.Next()
+}
+
+type IntersectionIterator struct {
+	aState iterState
+	bState iterState
+}
+
+func NewIntersectionIterator(iterA, iterB SetIterator) SetIterator {
+	d.Chk.NotNil(iterA)
+	d.Chk.NotNil(iterB)
+	a := iterState{i: iterA, v: iterA.Next()}
+	b := iterState{i: iterB, v: iterB.Next()}
+	return &IntersectionIterator{aState: a, bState: b}
+}
+
+func (u *IntersectionIterator) Next() Value {
+	for cont := true; cont; {
+		switch compareValue(u.aState.v, u.bState.v) {
+		case -1:
+			u.aState.AdvanceTo(u.bState.v)
+		case 0:
+			cont = false
+		case 1:
+			u.bState.AdvanceTo(u.aState.v)
+		}
+	}
+	// we only get here if aState and bState are equal
+	res := u.aState.v
+	u.aState.Next()
+	u.bState.Next()
+	return res
+}
+
+func (u *IntersectionIterator) AdvanceTo(v Value) Value {
+	d.Chk.NotNil(v)
+	if u.aState.v != nil && !v.Less(u.aState.v) {
```

Comment:
I've seen this pattern a few times now, maybe everything should just be using `compareValue` which both documents the behavior of nil, and does the check properly.

Up to you.